File: src/kernel/geometry-hydration.ts
import { normalizeGeometryStructure } from "../data/schema.js";
import { ProcessGeometryState } from "../process/process-geometry-state.js";

/**
 * Rebuild a mutable process geometry state from a stored geometry structure.
 */
export function geometryStructureToProcessGeometryState(payload: any, options: any = {}) {
  return ProcessGeometryState.fromStructure(payload, {
    footprint: { derive: "largestRootBody" },
    ...options,
  });
}

/**
 * Convert a ProcessGeometryState-like value back into a normalized structure.
 */
export function processGeometryStateToGeometryStructure(value: any) {
  if (
    value instanceof ProcessGeometryState ||
    (value?.toGeometryStructure && typeof value.toGeometryStructure === "function")
  ) {
    return value.toGeometryStructure();
  }
  return normalizeGeometryStructure(value);
}

File: packages/kernel-py/src/process_flow_kernel/serialization/geometry_hydration.py
from __future__ import annotations

from .schema import normalize_geometry_structure
from ..domain.process_geometry_state import ProcessGeometryState


def geometry_structure_to_process_geometry_state(payload, options=None):
    merged = {"footprint": {"derive": "largestRootBody"}}
    if options:
        merged.update(options)
    return ProcessGeometryState.from_structure(payload, merged)


def process_geometry_state_to_geometry_structure(value):
    if isinstance(value, ProcessGeometryState):
        return value.to_geometry_structure()
    if hasattr(value, "to_geometry_structure") and callable(value.to_geometry_structure):
        return value.to_geometry_structure()
    return normalize_geometry_structure(value)
